clippy_lints/src/indexing_slicing.rs

```diff
@@ -0,0 +1,201 @@
+//! lint on indexing and slicing operations
+
+use crate::consts::{constant, Constant};
+use crate::utils::higher::Range;
+use crate::utils::{self, higher};
+use rustc::hir::*;
+use rustc::lint::*;
+use rustc::ty;
+use syntax::ast::RangeLimits;
+
+/// **What it does:** Checks for out of bounds array indexing with a constant
+/// index.
+///
+/// **Why is this bad?** This will always panic at runtime.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = [1,2,3,4];
+///
+/// // Bad
+/// x[9];
+/// &x[2..9];
+///
+/// // Good
+/// x[0];
+/// x[3];
+/// ```
+declare_clippy_lint! {
+    pub OUT_OF_BOUNDS_INDEXING,
+    correctness,
+    "out of bounds constant indexing"
+}
+
+/// **What it does:** Checks for usage of indexing or slicing. Does not report
+/// if we can tell that the indexing or slicing operations on an array are in
+/// bounds.
+///
+/// **Why is this bad?** Indexing and slicing can panic at runtime and there are
+/// safe alternatives.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+/// ```rust
+/// let x = vec![0; 5];
+/// // Bad
+/// x[2];
+/// &x[2..100];
+/// &x[2..];
+/// &x[..100];
+///
+/// // Good
+/// x.get(2)
+/// x.get(2..100)
+/// x.get(2..)
+/// x.get(..100)
+/// ```
+declare_clippy_lint! {
+    pub INDEXING_SLICING,
+    pedantic,
+    "indexing/slicing usage"
+}
+
+#[derive(Copy, Clone)]
+pub struct IndexingSlicing;
+
+impl LintPass for IndexingSlicing {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INDEXING_SLICING, OUT_OF_BOUNDS_INDEXING)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IndexingSlicing {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if let ExprIndex(ref array, ref index) = &expr.node {
+            match &index.node {
+                // Both ExprStruct and ExprPath require this approach's checks
+                // on the `range` returned by `higher::range(cx, index)`.
+                // ExprStruct handles &x[n..m], &x[n..] and &x[..n].
+                // ExprPath handles &x[..] and x[var]
+                ExprStruct(..) | ExprPath(..) => {
+                    if let Some(range) = higher::range(cx, index) {
+                        let ty = cx.tables.expr_ty(array);
+                        if let ty::TyArray(_, s) = ty.sty {
+                            let size: u128 = s.assert_usize(cx.tcx).unwrap().into();
+                            // Index is a constant range.
+                            if let Some((start, end)) = to_const_range(cx, range, size) {
+                                if start > size || end > size {
+                                    utils::span_lint(
+                                        cx,
+                                        OUT_OF_BOUNDS_INDEXING,
+                                        expr.span,
+                                        "range is out of bounds",
+                                    );
```

Comment:
I would suggest to `return` after linting for an out of bounds range. On the one hand saying that the range is out of bounds here and then that it might panic with the `INDEXING_SLICING` lint is kind of redundant. On the other hand linting only once would be more consistent with the code below in the indexing `[n]` case.